pkg/handler/v2/external_cluster/kubeone.go

```diff
@@ -74,55 +80,163 @@ func importKubeOneCluster(ctx context.Context, name string, userInfoGetter func(
+	manifest := kubeOneSpec.ManifestReference
+
+	if kubeOneSpec.ClusterStatus.Status == kubermaticv1.StatusReconciling {
+		return nil, errors.NewBadRequest("Operation is not allowed: Another operation: (Upgrading) is in progress, please wait for it to finish before starting a new operation.")
+	}
+
+	containerRuntimeDocker, err := isContainerRuntimeDocker(ctx, externalCluster, externalClusterProvider)
+	if err != nil {
+		return nil, err
+	}
+
+	if !containerRuntimeDocker {
+		return nil, nil
+	}
+
 	manifestSecret := &corev1.Secret{}
 	if err := masterClient.Get(ctx, types.NamespacedName{Namespace: manifest.Namespace, Name: manifest.Name}, manifestSecret); err != nil {
 		return nil, errors.NewBadRequest(fmt.Sprintf("can not retrieve kubeone manifest secret: %v", err))
 	}
-	oldManifestSecret := manifestSecret.DeepCopy()
+	manifestValue := manifestSecret.Data[resources.KubeOneManifest]
+	kubeOneClusterManifest := &kubeonev1beta2.KubeOneCluster{}
+	if err := yaml.UnmarshalStrict(manifestValue, kubeOneClusterManifest); err != nil {
+		return nil, err
+	}
+	kubeOneClusterManifest.ContainerRuntime.Docker = nil
+	kubeOneClusterManifest.ContainerRuntime.Containerd = &kubeonev1beta2.ContainerRuntimeContainerd{}
+	yamlManifest, err := yaml.Marshal(kubeOneClusterManifest)
+	if err != nil {
+		return nil, err
+	}
 
+	oldManifestSecret := manifestSecret.DeepCopy()
 	manifestSecret.Data = map[string][]byte{
 		resources.KubeOneManifest: yamlManifest,
 	}
 	if err := masterClient.Patch(ctx, manifestSecret, ctrlruntimeclient.MergeFrom(oldManifestSecret)); err != nil {
-		return nil, fmt.Errorf("failed to update kubeone manifest secret for upgrade version %s/%s: %w", manifest.Name, manifest.Namespace, err)
+		return nil, fmt.Errorf("failed to update kubeone manifest secret for container-runtime containerd %s/%s: %w", manifest.Name, manifest.Namespace, err)
 	}
 
 	oldexternalCluster := externalCluster.DeepCopy()
 	// update kubeone externalcluster status.
 	externalCluster.Spec.CloudSpec.KubeOne.ClusterStatus.Status = kubermaticv1.StatusReconciling
-	if upgradeMD == "true" {
-		externalCluster.Spec.CloudSpec.KubeOne.ClusterStatus.StatusMessage = "upgrading machine MD"
-	}
-	// update api externalcluster status
-	newCluster.Status.State = apiv2.RECONCILING
+	// update api externalcluster status.
+	newCluster.Status = apiv2.ExternalClusterStatus{State: apiv2.RECONCILING}
+
 	if err := masterClient.Patch(ctx, externalCluster, ctrlruntimeclient.MergeFrom(oldexternalCluster)); err != nil {
 		return nil, fmt.Errorf("failed to update kubeone cluster status %s: %w", externalCluster.Name, err)
 	}
 
 	return newCluster, nil
 }
+
+func isContainerRuntimeDocker(ctx context.Context,
+	externalCluster *kubermaticv1.ExternalCluster,
+	externalClusterProvider provider.ExternalClusterProvider,
+) (bool, error) {
+	var containerRuntimeDocker bool
+	nodes, err := externalClusterProvider.ListNodes(ctx, externalCluster)
+	if err != nil {
+		return containerRuntimeDocker, err
```

Comment:
this one is dangerous. You don't know what kind of error happened. You should retrieve the container runtime type not assume it's docker. If ListNodes throws an error because of controller runtime then you should have some predefined error for this. You can have a runtime error, connection error, and even more

Reply:
good point! thanks:)